### ISSUE TITLE
ha: anti-affnity rules are too restrictive

### ISSUE THREAD

**Issue by olix0r** [linkerd/linkerd2#3939]
Linkerd, installed with `--ha` on a 3-node cluster, cannot be successfully upgraded:

```
:; kubectl get events -n linkerd
LAST SEEN   TYPE      REASON             OBJECT                                        MESSAGE
94s         Warning   FailedScheduling   pod/linkerd-controller-c845b4549-wbz2d        0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.
94s         Warning   FailedScheduling   pod/linkerd-destination-6488b657b-5srp8       0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.
94s         Warning   FailedScheduling   pod/linkerd-identity-7dc5d97cf9-rxnz2         0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.
94s         Warning   FailedScheduling   pod/linkerd-proxy-injector-64d8d46794-lb69h   0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.
94s         Warning   FailedScheduling   pod/linkerd-sp-validator-86bbcbfbff-sldx6     0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.
94s         Warning   FailedScheduling   pod/linkerd-tap-c895b586d-ps22z               0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.
```

@grampelberg suggests that this can be resolved with the following:

- define a PDB - https://kubernetes.io/docs/tasks/run-application/configure-pdb/#specifying-a-poddisruptionbudget
- set progressDeadlineSeconds
- define maxUnavailable
- configure maxSurge

**Comment by kleimkuhler** [comment]
@grampelberg I've been looking into this this morning and I'm not sure a PDB is the answer here. I think we either need to relax the anti-affinity rules from require to prefer so that pods can be scheduled on a node that has other control plane pods, or recommend HA deployments on at least 4 node clusters.

Using a simpler deployment of nginx with three replicas and the same Linkerd anti-affinity rules on a three node cluster, I can reproduce the deployment getting stuck spinning up pods even with a PDB applied.

I tested PDBs with `MaxUnavailable: 1` as well as `MinAvailable: 1`. In both cases, the new pod gets stuck with the status message: `0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't satisfy existing pods anti-affinity rules.`

PDB has not been working because once the deployment has been changed such that a restart is required, k8s does not seem to terminate the existing pod(s) before starting new ones. The new one then cannot fit on a node and the cluster gets stuck. The docs suggest that this is a "disruption" and therefore should all a pod to be terminated to make room for another, but that has not been the case.

**Comment by grampelberg** [comment]
@kleimkuhler that all sounds pretty reasonable to me. Mind sharing the nginx config you're testing with?

@siggy and I found instances where 5 nodes were required, but only for one of our components (and only sometimes).

If we're doing what the docs say to do, feels like the next step would be to find/open a bug on k8s itself and then reference that here. That way folks will at least understand what the deal is.

**Comment by kleimkuhler** [comment]
@grampelberg This config is what replicates the issue we are seeing. There is a PDB at the bottom with `MinAvailable: 1` (but the issue continues to exist with `MaxUnavailable: 2`:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
    foo: bar
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchExpressions:
                    - key: app
                      operator: In
                      values:
                        - nginx
                topologyKey: failure-domain.beta.kubernetes.io/zone
              weight: 100
          requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                  - key: app
                    operator: In
                    values:
                      - nginx
              topologyKey: kubernetes.io/hostname
      containers:
        - name: nginx
          image: nginx:1.14.2
          ports:
            - containerPort: 80
---
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
    name: nginx-pdb
    labels:
        app: nginx
spec:
    minAvailable: 1
    selector:
        matchLabels:
            app: nginx
```

I have successfully upgraded Linkerd a few times now on a 4 node GKE cluster, so I haven't observed the need for 5 nodes, but maybe Siggy has seen that more recently?

**Comment by kleimkuhler** [comment]
#4260 should fix the issue where upgrade fails on a 4 node cluster.

**Comment by kleimkuhler** [comment]
Reported here: kubernetes/kubernetes#90151

**Comment by kleimkuhler** [comment]
Closed by #4267
